airflow/api_fastapi/execution_api/routes/task_instances.py

```diff
@@ -393,6 +394,42 @@ def ti_put_rtif(
     return {"message": "Rendered task instance fields successfully set"}
 
 
+@router.get(
+    "/{task_instance_id}/previous_successful_dagrun",
+    status_code=status.HTTP_200_OK,
+    responses={
+        status.HTTP_404_NOT_FOUND: {"description": "Task Instance or Dag Run not found"},
+    },
+)
+def get_previous_successful_dagrun(
+    task_instance_id: UUID, session: SessionDep
+) -> PrevSuccessfulDagRunResponse:
+    """
+    Get the previous successful DagRun for a TaskInstance.
+
+    The data from this endpoint is used to get values for Task Context.
+    """
+    ti_id_str = str(task_instance_id)
+    task_instance = session.scalar(select(TI).where(TI.id == ti_id_str))
+    if not task_instance:
+        return PrevSuccessfulDagRunResponse()
+
+    dag_run = session.scalar(
+        select(DR)
+        .where(
+            DR.dag_id == task_instance.dag_id,
+            DR.logical_date < task_instance.logical_date,
```

Comment:
I think @uranusjr is about to break this. Quick, lets merge it while it still works >_>